server/sonar-server/src/main/java/org/sonar/server/metric/ws/CreateAction.java

```diff
@@ -0,0 +1,176 @@
+import org.sonar.api.server.ws.Response;
+import org.sonar.api.server.ws.WebService;
+import org.sonar.api.utils.text.JsonWriter;
+import org.sonar.core.metric.db.MetricDto;
+import org.sonar.core.permission.GlobalPermissions;
+import org.sonar.core.persistence.DbSession;
+import org.sonar.core.persistence.MyBatis;
+import org.sonar.server.db.DbClient;
+import org.sonar.server.exceptions.ServerException;
+import org.sonar.server.user.UserSession;
+
+public class CreateAction implements MetricsWsAction {
+  private static final String ACTION = "create";
+  public static final String PARAM_NAME = "name";
+  public static final String PARAM_KEY = "key";
+  public static final String PARAM_TYPE = "type";
+  public static final String PARAM_DESCRIPTION = "description";
+  public static final String PARAM_DOMAIN = "domain";
+
+  private static final String FIELD_ID = "id";
+  private static final String FIELD_NAME = PARAM_NAME;
+  private static final String FIELD_KEY = PARAM_KEY;
+  private static final String FIELD_TYPE = PARAM_TYPE;
+  private static final String FIELD_DESCRIPTION = PARAM_DESCRIPTION;
+  private static final String FIELD_DOMAIN = PARAM_DOMAIN;
+
+  private final DbClient dbClient;
+  private final UserSession userSession;
+
+  public CreateAction(DbClient dbClient, UserSession userSession) {
+    this.dbClient = dbClient;
+    this.userSession = userSession;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction(ACTION)
+      .setPost(true)
+      .setDescription("Create custom metric.<br /> Requires 'Administer System' permission.")
+      .setSince("5.2")
+      .setHandler(this);
+
+    action.createParam(PARAM_NAME)
+      .setRequired(true)
+      .setDescription("Name")
+      .setExampleValue("Team Size");
+
+    action.createParam(PARAM_KEY)
+      .setRequired(true)
+      .setDescription("Key")
+      .setExampleValue("team_size");
+
+    action.createParam(PARAM_TYPE)
+      .setRequired(true)
+      .setDescription("Metric type key")
+      .setPossibleValues(Metric.ValueType.names())
+      .setExampleValue(String.format("<ul><li>%s</li><li>%s</li></ul>", Metric.ValueType.INT.name(), Metric.ValueType.BOOL.name()));
+
+    action.createParam(PARAM_DESCRIPTION)
+      .setDescription("Description")
+      .setExampleValue("Size of the team");
+
+    action.createParam(PARAM_DOMAIN)
+      .setDescription("Domain")
+      .setExampleValue("Tests");
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    userSession.checkLoggedIn().checkGlobalPermission(GlobalPermissions.SYSTEM_ADMIN);
+    String key = request.mandatoryParam(PARAM_KEY);
+
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      MetricDto metricInDb = dbClient.metricDao().selectNullableByKey(dbSession, key);
+      checkMetricInDb(metricInDb, key);
+
+      MetricDto metricTemplate = new MetricDto()
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Possible null pointer dereference in org.sonar.server.metric.ws.CreateAction.handle(Request, Response) due to return value of called method [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE)